File: 2024/Python/Day8/solutionA.py
# parse raw data into dictionary
def parse_data(data):
    rows_data = data.split("\n")[:-1]
    cols_data = list(data.split("\n")[0])
    parsed = {}
    for row in range(len(rows_data)):
        for col in range(len(cols_data)):
            cell = rows_data[row][col]
            if cell != ".":
                if cell not in parsed:
                    parsed[cell] = []
                parsed[cell].append((row, col))

    return parsed

# check if the cell is within the bounds of the grid
def in_bounds(r, c, rows, cols):
    return r >= 0 and r < rows and c >= 0 and c < cols

# count the number of antinodes
def count_antinodes(data):
    rows = len(data.split("\n")[:-1])
    cols = len(list(data.split("\n")[0]))
    antennas = parse_data(data)
    antinodes = []

    for antenna in antennas:
        length = len(antennas[antenna])

        for i in range(length):
            for j in range(length):
                if i == j:
                    continue
                
                (r1, c1) = antennas[antenna][i]
                (r2, c2) = antennas[antenna][j]

                dr = r2 - r1
                dc = c2 - c1

                r3 = r1 - dr 
                c3 = c1 - dc 

                r4 = r2 + dr 
                c4 = c2 + dc 

                if in_bounds(r3, c3, rows, cols):
                    antinodes.append(f"{r3},{c3}")

                if in_bounds(r4, c4, rows, cols):
                    antinodes.append(f"{r4},{c4}")

    return len(set(antinodes))

data = ""
with open("../../data/day8.txt") as file:
    for line in file:
        data += line

print(count_antinodes(data))

File: 2024/JS/Day8/solutionA.js
const fs = require("fs");

function countAntinodes(mapInput) {
  const map = mapInput
    .split("\n")
    .slice(0, -1)
    .map((line) => line.split(""));
  const rows = map.length;
  const cols = map[0].length;
  const antennas = {};

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      const cell = map[row][col];
      if (cell !== ".") {
        if (!antennas[cell]) {
          antennas[cell] = [];
        }
        antennas[cell].push([row, col]);
      }
    }
  }

  const antinodeSet = new Set();

  for (const [frequency, positions] of Object.entries(antennas)) {
    const len = positions.length;

    for (let i = 0; i < len; i++) {
      for (let j = 0; j < len; j++) {
        if (i === j) continue;

        const [r1, c1] = positions[i];
        const [r2, c2] = positions[j];

        const dr = r2 - r1;
        const dc = c2 - c1;

        const r3 = r1 - dr;
        const c3 = c1 - dc;
        const r4 = r2 + dr;
        const c4 = c2 + dc;

        // Add antinodes if they are within bounds
        if (r3 >= 0 && r3 < rows && c3 >= 0 && c3 < cols) {
          antinodeSet.add(`${r3},${c3}`);
        }
        if (r4 >= 0 && r4 < rows && c4 >= 0 && c4 < cols) {
          antinodeSet.add(`${r4},${c4}`);
        }
      }
    }
  }

  return antinodeSet.size;
}

fs.readFile("../../data/day8.txt", (err, data) => {
  let input = data.toString();
  console.log(countAntinodes(input));
});
